### ISSUE TITLE
"unlinkat /workspace/node_modules/end-of-stream/package.json: permission denied" after specifying "include" files

### ISSUE THREAD

**Issue by bervol** [buildpacks/pack#987]
### Summary
I use `pack build party-service --builder paketobuildpacks/builder:base` to build my NodeJS project.

When I do not specifically include the files for the build, the build is working.

However, when I add this `project.toml` file:
```toml
[build]
include = [
    "src",
    "package.json",
    "package-lock.json",
    "nest-cli.json",
    "tsconfig.*json",
]
```

The build fails with

> unlinkat /workspace/node_modules/end-of-stream/package.json: permission denied

---

##### Steps

1. Add above `project.toml` file.
2. `pack build party-service --builder paketobuildpacks/builder:base`

##### Current behavior
The build fails with

> unlinkat /workspace/node_modules/end-of-stream/package.json: permission denied

##### Expected behavior
Build should run successfully.

---

##### pack info
Pack:
  Version:  0.15.1+git-79adc30.build-1660
  OS/Arch:  darwin/amd64

Default Lifecycle Version:  0.9.3

Supported Platform APIs:  0.3, 0.4

Config:
(no config file found at /Users/bernhard.vollnhofer/.pack/config.toml)

##### docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.4.2-docker)
  scan: Docker Scan (Docker Inc., v0.5.0)

Server:
 Containers: 3
  Running: 0
  Paused: 0
  Stopped: 3
 Images: 12
 Server Version: 20.10.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.121-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 6
 Total Memory: 1.942GiB
 Name: docker-desktop
 ID: HLCK:QQSZ:NFKQ:CGOJ:JO36:A2PT:DJGD:TDPQ:5QS4:SMSG:44QA:PXGB
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

**Comment by bernhardvollnhofer** [comment]
I also get the error with the builder `gcr.io/buildpacks/builder:v1`:

> Failure: (ID: 2982a97d) removing node_modules: unlinkat ./node_modules/end-of-stream/package.json: permission denied

**Comment by dfreilich** [comment]
Hey @bernhardvollnhofer , thanks for opening the issue. I tried to replicate this, using the [paketo-buildpacks/samples](https://github.com/paketo-buildpacks/samples) repo, but I was unable to. Could you point me to a sample app, or the repo where you're seeing this issue? I'm assuming that this is a problem with the buildpacks, rather than with pack, but I can't pin that down without a reproduction of the issue.

**Comment by bernhardvollnhofer** [comment]
> Hey @bernhardvollnhofer , thanks for opening the issue. I tried to replicate this, using the [paketo-buildpacks/samples](https://github.com/paketo-buildpacks/samples) repo, but I was unable to. Could you point me to a sample app, or the repo where you're seeing this issue? I'm assuming that this is a problem with the buildpacks, rather than with pack, but I can't pin that down without a reproduction of the issue.

Thank you for your reply.

I have created a public repository with which this issue can be easily reproduced: https://github.com/bernhardvollnhofer/buildpack-nodejs-poc

Greetings

**Comment by bernhardvollnhofer** [comment]
When I use 
```bash
pack build buildpack-nodejs-poc --builder gcr.io/buildpacks/builder:v1
```
I am getting the same error:
```
Running "bash -c curl --fail --show-error --silent --location --retry 3 https://nodejs.org/dist/v14.15.3/node-v14.15.3-linux-x64.tar.xz | tar xJ --directory /layers/google.nodejs.runtime/node --strip-components=1"
Done "bash -c curl --fail --show-error --silent --location --retry..." (3.61205398s)
=== Node.js - NPM (google.nodejs.npm@0.9.0) ===
Failure: (ID: 2982a97d) removing node_modules: unlinkat ./node_modules/end-of-stream/package.json: permission denied
```
So I think it is related to `pack`, and not to the builder.

**Comment by dfreilich** [comment]
Those logs are in the `BUILDING` phase, when pack just runs the buildpacks, so I don't think this is an issue with `pack`. Further, I tried replicating the issue with that sample app, and I couldn't get it:
```
$ cat project.toml
[build]
include = [
    "src",
    "package.json",
    "package-lock.json",
    "nest-cli.json",
    "tsconfig.*json",
]
```

```
$  pack build buildpack-nodejs-poc --builder gcr.io/buildpacks/builder:v1
...
Status: Downloaded newer image for gcr.io/buildpacks/gcp/run:v1
===> DETECTING
3 of 5 buildpacks participating
google.nodejs.runtime 0.9.2
google.nodejs.npm     0.9.0
google.utils.label    0.0.1
===> ANALYZING
Previous image with name "buildpack-nodejs-poc" not found
===> RESTORING
===> BUILDING
=== Node.js - Runtime (google.nodejs.runtime@0.9.2) ===
Resolving Node.js version based on semver "14.x.x"
--------------------------------------------------------------------------------
Running "curl --fail --show-error --silent --location --get --data-urlencode range=14.x.x http://semver.io/node/resolve"
14.15.3Done "curl --fail --show-error --silent --location --get --data-ur..." (538.1614ms)
Using resolved runtime version from package.json: 14.15.3
Installing Node.js v14.15.3
--------------------------------------------------------------------------------
Running "bash -c curl --fail --show-error --silent --location --retry 3 https://nodejs.org/dist/v14.15.3/node-v14.15.3-linux-x64.tar.xz | tar xJ --directory /layers/google.nodejs.runtime/node --strip-components=1"
Done "bash -c curl --fail --show-error --silent --location --retry..." (4.1863204s)
=== Node.js - NPM (google.nodejs.npm@0.9.0) ===
Installing application dependencies.
--------------------------------------------------------------------------------
Running "npm ci --quiet (NODE_ENV=production)"

> @nestjs/core@7.6.4 postinstall /workspace/node_modules/@nestjs/core
> opencollective || exit 0

added 106 packages in 12.799s
Done "npm ci --quiet (NODE_ENV=production)" (13.5413205s)
=== Utils - Label Image (google.utils.label@0.0.1) ===
===> EXPORTING
Adding layer 'google.nodejs.runtime:node'
Adding layer 'google.nodejs.npm:env'
Adding 1/1 app layer(s)
Adding layer 'launcher'
Adding layer 'config'
Adding layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
*** Images (91b1dc6907e9):
      buildpack-nodejs-poc
Adding cache layer 'google.nodejs.runtime:node'
Adding cache layer 'google.nodejs.npm:npm'
Successfully built image buildpack-nodejs-poc
```

```
$  pack build buildpack-nodejs-poc2 --builder paketobuildpacks/builder:base --clear-cache
base: Pulling from paketobuildpacks/builder
Digest: sha256:984a3684db80a6d53214b81a9f21c31529bede5b447d6d6d82d94cd6734d2424
Status: Image is up to date for paketobuildpacks/builder:base
base-cnb: Pulling from paketobuildpacks/run
Digest: sha256:f393fa2927a2619a10fc09bb109f822d20df909c10fed4ce3c36fad313ea18e3
Status: Image is up to date for paketobuildpacks/run:base-cnb
===> DETECTING
paketo-buildpacks/node-engine 0.1.4
paketo-buildpacks/npm-install 0.2.3
paketo-buildpacks/npm-start   0.0.3
===> ANALYZING
Previous image with name "buildpack-nodejs-poc2" not found
Skipping buildpack layer analysis
===> BUILDING
Paketo Node Engine Buildpack 0.1.4
  Resolving Node Engine version
    Candidate version sources (in priority order):
                -> ""
      <unknown> -> "*"

    Selected Node Engine version (using ): 14.15.1

  Executing build process
    Installing Node Engine 14.15.1
      Completed in 7.026s

  Configuring environment
    NODE_ENV     -> "production"
    NODE_HOME    -> "/layers/paketo-buildpacks_node-engine/node"
    NODE_VERBOSE -> "false"

    Writing profile.d/0_memory_available.sh
      Calculates available memory based on container limits at launch time.
      Made available in the MEMORY_AVAILABLE environment variable.

Paketo NPM Install Buildpack 0.2.3
  Resolving installation process
    Process inputs:
      node_modules      -> "Not found"
      npm-cache         -> "Not found"
      package-lock.json -> "Found"

    Selected NPM build process: 'npm ci'

  Executing build process
    Running 'npm ci --unsafe-perm --cache /layers/paketo-buildpacks_npm-install/npm-cache'
      Completed in 4.47s

  Configuring environment
    NPM_CONFIG_LOGLEVEL   -> "error"
    NPM_CONFIG_PRODUCTION -> "true"
    PATH                  -> "$PATH:/layers/paketo-buildpacks_npm-install/modules/node_modules/.bin"

Paketo NPM Start Buildpack 0.0.3
  Assigning launch processes
    web: nest start
===> EXPORTING
Adding layer 'paketo-buildpacks/node-engine:node'
Adding layer 'paketo-buildpacks/npm-install:modules'
Adding layer 'paketo-buildpacks/npm-install:npm-cache'
Adding 1/1 app layer(s)
Adding layer 'launcher'
Adding layer 'config'
Adding layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
*** Images (3d80034d2610):
      buildpack-nodejs-poc2
Adding cache layer 'paketo-buildpacks/node-engine:node'
Adding cache layer 'paketo-buildpacks/npm-install:modules'
Adding cache layer 'paketo-buildpacks/npm-install:npm-cache'
Successfully built image buildpack-nodejs-poc2
```

I'm running the same version of `pack` as you. Can you try it with the heroku builder as well?

**Comment by bvollnhofer** [comment]
A colleague of mine has the same error. We are both using macOS 10.15.7. Could it be that it is related to macOS?

**Comment by dfreilich** [comment]
@bvollnhofer I'm also on MacOS 10.15.7, as is most of the development team, so I don't think it's related to that.

**Comment by ekcasey** [comment]
@dfreilich I could produce a slightly different issue by installing the node modules on my host machine before building the paketo sample app with the provided `project.toml`. It looks like when `pack` is used with `build.include` the non-included files are still copied into the container?

**Comment by dfreilich** [comment]
Thanks for the help there @ekcasey , I was able to replicate it as well, once I ran a local `npm install`, with both the google and paketo builder.

**Comment by dwillist** [comment]
So this issue should be fixed by #1043, and some docs updates to talk a bit about how `gitignore` syntax works. The  `project.toml` file included in this issue will need to be changed a bit due to how `gitignore` syntax works. I think you probably want something like this:

```
[build]
include = [
    "/src",
    "/package.json",
    "/package-lock.json",
    "/nest-cli.json",
    "/tsconfig.*json",
]
```
so that ONLY top level files are included. The `package.json` entry without a preceding `/` was causing issues because you end up including all `package.json` files nested in the `node_modules` folder. Hench the error `removing node_modules: unlinkat ./node_modules/end-of-stream/package.json: permission `

**Comment by dwillist** [comment]
Closing this issue as fix is merged
